components/search/src/filter.rs

```diff
@@ -241,6 +266,49 @@ fn find_engine_with_match(
         .map(|e| e.identifier.clone())
 }
 
+fn find_engine_with_match_mut<'a>(
+    engines: &'a mut [SearchEngineDefinition],
+    engine_id_match: &String,
+) -> Option<&'a mut SearchEngineDefinition> {
+    if engine_id_match.is_empty() {
+        return None;
+    }
+    if let Some(match_no_star) = engine_id_match.strip_suffix('*') {
+        return engines
+            .iter_mut()
+            .find(|e| e.identifier.starts_with(match_no_star))
+    }
+
+    engines
+        .iter_mut()
+        .find(|e| e.identifier == *engine_id_match)
+}
+
+fn get_priority(
+    engine: &SearchEngineDefinition,
+    default_engine_id: Option<&String>,
+    default_private_engine_id: Option<&String>,
+    ) -> u32 {
+    if Some(&engine.identifier) == default_engine_id {
+        return u32::MAX;
+    }
+    if Some(&engine.identifier) == default_private_engine_id {
+        return u32::MAX - 1;
+    }
+    engine.order_hint.unwrap_or(0)
+}
+
+fn sort(
+    default_engine_id: Option<&String>,
+    default_private_engine_id: Option<&String>,
+    a: &SearchEngineDefinition,
+    b: &SearchEngineDefinition
+) -> std::cmp::Ordering {
+    let b_index = get_priority(b, default_engine_id, default_private_engine_id);
+    let a_index = get_priority(a, default_engine_id, default_private_engine_id);
+    b_index.cmp(&a_index)
```

Comment:
There's a piece missing here, which the engine selector on desktop doesn't do - the items which aren't default/have an order hint set, need to be sorted alphabetically by their display name.

This is the full function we use: https://searchfox.org/mozilla-central/rev/5b061cdc4d40d44988dc61aa941cfbd98e31791f/toolkit/components/search/SearchUtils.sys.mjs#362-440

One thing I'm not sure about is how we'll handle the locale-based sorting like we do on desktop. I haven't looked into the options for sort functions on Rust.